### PR TITLE
Add ArraySegment encode/decode

### PR DESCRIPTION
**Pull request by 7sharp9**
Tests in progress...
Any comments so far @gusty ?

### CODE REVIEW

**Comment by gusty** [comment]
Why do you need to special-case the netcoreapp ?

**Comment by 7sharp9** [comment]
@gusty `ToArray` is missing on `ArraySegment` for non netcoreapp versions

File: Tests/Tests.fs [hunk 1]
```diff
@@ -162,6 +162,18 @@ type Assert with
...
+  static member ArraySegment() =
+      {new Arbitrary<ArraySegment<int>>() with
+          override x.Generator =
+              let generator =
+                gen {
+                    let! values = Arb.generate
+                    return values |> ArraySegment<int>
+                }
+              generator
+          override x.Shrinker t = Seq.empty }
```
Comment (wallymathieu) [inline Tests/Tests.fs]: this looks a bit weird, shouldn't you adjust based on t?

Comment (wallymathieu) [reply Tests/Tests.fs]: Perhaps you could use Default.Array arbitrary as base?

Comment (wallymathieu) [reply Tests/Tests.fs]: I tried changing to 
```
type ArraySegmentGenerator =
  static member ArraySegment() =
            Arb.Default.Array()
            |> Arb.convert ArraySegment<int> (fun (s:ArraySegment<int>) -> s.ToArray())
```
what goes wrong is that it seems that it tries to check if a value wrapped by result is equal to a value not wrapped by result.

File: Tests/Tests.fs [hunk 2]
```diff
@@ -494,6 +510,7 @@ let tests = [
             yield testProperty "string list" (roundtrip<string list>)
             yield testProperty "string set" (roundtrip<string Set>)
             yield testProperty "int array" (roundtrip<int array>)
+            yield testProperty "int ArraySegment" (roundtrip<int ArraySegment> )
```
Comment (wallymathieu) [inline Tests/Tests.fs]:
```
            let eqArraySegment (a:ArraySegment<_>) (b:ArraySegment<_>) = a.ToArray() = b.ToArray()
```

then
```
            yield testProperty "int ArraySegment" (roundtripEq<int ArraySegment> eqArraySegment)
```

Comment (wallymathieu) [reply Tests/Tests.fs]: perhaps more clearly as:
``` yield testProperty "int ArraySegment" (roundtripEq<int ArraySegment> (Seq.forall2 (=))) ```

**Pull request merged**